lib/gssapi/sanon/accept_sec_context.c

```diff
@@ -73,30 +75,50 @@ _gss_sanon_accept_sec_context(OM_uint32 *minor,
     if (major != GSS_S_COMPLETE)
 	goto out;
 
+    sc = calloc(1, sizeof(*sc));
     if (sc == NULL) {
-	sc = calloc(1, sizeof(*sc));
-	if (sc == NULL) {
-	    *minor = ENOMEM;
-	    major = GSS_S_FAILURE;
-	    goto out;
-	}
+	*minor = ENOMEM;
+	major = GSS_S_FAILURE;
+	goto out;
+    }
+
+    /* initiator token can include optional 64-bit flags */
+    if (mech_input_token.length != crypto_scalarmult_curve25519_BYTES &&
+	mech_input_token.length != crypto_scalarmult_curve25519_BYTES + 8) {
+	*minor = 0;
+	major = GSS_S_DEFECTIVE_TOKEN;
+	goto out;
     }
 
+    initiator_pk = mech_input_token;
+    initiator_pk.length = crypto_scalarmult_curve25519_BYTES;
+
     /* compute public and secret keys */
     major = _gss_sanon_curve25519_base(minor, sc);
     if (major != GSS_S_COMPLETE)
 	goto out;
 
+    if (mech_input_token.length > crypto_scalarmult_curve25519_BYTES) {
+	/* extra flags */
+	uint8_t *p = (uint8_t *)mech_input_token.value + crypto_scalarmult_curve25519_BYTES;
+	uint32_t dummy;
+
+	_gss_mg_decode_be_uint32(p, &dummy); /* higher 64 bits not used yet */
+	_gss_mg_decode_be_uint32(&p[4], &req_flags);
+    }
+
+    req_flags &= SANON_PROTOCOL_FLAG_MASK; /* do not let initiator set any other flags */
+
     /* compute shared secret */
-    major = _gss_sanon_curve25519(minor, sc, &mech_input_token,
+    major = _gss_sanon_curve25519(minor, sc, &initiator_pk, req_flags,
 				  input_chan_bindings, &session_key);
     if (major != GSS_S_COMPLETE)
 	goto out;
 
-    sc->flags |= GSS_C_REPLAY_FLAG | GSS_C_SEQUENCE_FLAG | GSS_C_CONF_FLAG |
```

Comment:
Ahh, we no longer need `sc->flags` because they go into the RFC4121 context, got it.